src/storage/src/sink/kafka/util.rs

```diff
@@ -389,12 +391,19 @@ pub struct ProgressRecord {
 
 /// Determines the latest progress record from the specified topic for the given
 /// key (e.g. akin to a sink's GlobalId).
-pub async fn determine_latest_progress_record(
+pub(super) async fn determine_latest_progress_record(
     name: String,
     progress_topic: String,
     progress_key: String,
     progress_client: Arc<BaseConsumer<TunnelingClientContext<MzClientContext>>>,
+    producer: &KafkaTxProducer,
 ) -> Result<Option<Timestamp>, anyhow::Error> {
+    // Ensure that the producer has initialized transactions and is the current
+    // holder of its transaction ID. We need this to ensure that we get the most
+    // recent offsets from the progress topic.
+    producer.begin_transaction().await?;
+    producer.abort_transaction().await?;
```

Comment:
Wait, I thought `init_transactions()` was what fenced out old versions of the producer?

Reply:
This is ensuring that `init_transactions` was called; will adjust the note here

Reply:
Can we enforce that through types instead? It seems like `KafkaTxProducer` could plausibly be the place where we uphold that invariant. Like, stick that type in its own module so the fields are private, and then force construction through an `async fn new() -> Result<..., _>` that forces a call to `init_transaction()` before returning.

Reply:
Sure thing